src/app/codegen-data-model/CodegenDataModel_Write.cpp

```diff
@@ -0,0 +1,381 @@
+                            CHIP_IM_GLOBAL_STATUS(NeedsTimedInteraction));
+    }
+
+    // Extra check: internal requests can bypass the read only check, however global attributes
+    // have no underlying storage, so write still cannot be done
+    VerifyOrReturnError(attributeMetadata != nullptr, CHIP_IM_GLOBAL_STATUS(UnsupportedWrite));
+
+    if (request.path.mDataVersion.HasValue())
+    {
+        std::optional<InteractionModel::ClusterInfo> clusterInfo = GetClusterInfo(request.path);
+        if (!clusterInfo.has_value())
+        {
+            ChipLogError(DataManagement, "Unable to get cluster info for Endpoint 0x%x, Cluster " ChipLogFormatMEI,
+                         request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
+        }
+
+        if (request.path.mDataVersion.Value() != clusterInfo->dataVersion)
+        {
+            ChipLogError(DataManagement, "Write Version mismatch for Endpoint 0x%x, Cluster " ChipLogFormatMEI,
+                         request.path.mEndpointId, ChipLogValueMEI(request.path.mClusterId));
+            return CHIP_IM_GLOBAL_STATUS(DataVersionMismatch);
+        }
+    }
+
+    AttributeAccessInterface * aai       = GetAttributeAccessOverride(request.path.mEndpointId, request.path.mClusterId);
+    std::optional<CHIP_ERROR> aai_result = TryWriteViaAccessInterface(request.path, aai, decoder);
+    if (aai_result.has_value())
+    {
+        return *aai_result;
+    }
+
+    ReturnErrorOnFailure(DecodeValueIntoEmberBuffer(decoder, *attributeMetadata, gEmberAttributeIOBufferSpan));
+
+    Protocols::InteractionModel::Status status;
+
+    if (!request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+
+        EmberAfAttributeSearchRecord record;
+        record.endpoint    = request.path.mEndpointId;
+        record.clusterId   = request.path.mClusterId;
+        record.attributeId = request.path.mAttributeId;
+        // NOTE: gEmberAttributeIOBufferSpah is likely much larger than the attribute itself. Ember is supposed
+        //       to be able to pick up the "valid size" out of this
+        //
+        // Specifically `When writing attributes, readLength is ignored.`.
+        // The actual length is used by unit tests to copy over data without needing "get attribute size" logic
+        // since metadata attribute sizes in our mocks is not generally correct.
+        status = emAfReadOrWriteAttribute(&record, attributeMetadata, /* buffer = */ gEmberAttributeIOBufferSpan.data(),
+                                          /* readLength = */ static_cast<uint16_t>(gEmberAttributeIOBufferSpan.size()),
+                                          /* write = */ true);
+    }
+    else
+    {
+        status = emAfWriteAttributeExternal(request.path.mEndpointId, request.path.mClusterId, request.path.mAttributeId,
```

Comment:
Again, what happened to the "is the data we are writing larger than attributeMetadata->size" check?  emAfWriteAttributeExternal can't do that check, because it does not have the right information (e.g. the size of the data we are writing).  This will silently truncate too-long strings instead of correctly erroring out, as far as I can tell.

Reply:
Added that one back with a unit test.

This was really painful as our unit test fixtures did not really set sizes for attributes, however I do want the external paths in unit tests because of ACL and writability tests.